// sbis3-ws/Core/I18n.js
define('Core/I18n', [
    'require',
    'exports',
    'Core/Util'
], function (require, exports, Util_1) {
    'use strict';
    Object.defineProperty(exports, '__esModule', { value: true });
    var availableLanguage = {
        'ru-RU': 'Русский (Россия)',
        'en-US': 'English (USA)'
    };
    var global = function () {
            return this || (0, eval)('this');
        }(), localizationEnabled = Util_1.constants.isServerScript ? false : Util_1.constants.isNodePlatform ? true : global.contents ? !!global.contents.defaultLanguage : Util_1.constants.i18n, directoryRegexp = /(.*\/)?(resources\/(.+?)|ws)\//, dblSlashes = /\\/g, i18n, PLURAL_PREFIX = 'plural#', PLURAL_DELIMITER = '|', EXPIRES_COOKIES = 2920;
    var RkString = function RkString(value, resolver) {
        Object.defineProperties(this, {
            translatedValue: {
                enumerable: true,
                get: function () {
                    return String(resolver(value) || value);
                }
            },
            length: {
                enumerable: true,
                get: function () {
                    return this.translatedValue.length;
                }
            }
        });
    };
    RkString.prototype = Object.create(String.prototype);
    RkString.prototype.toString = RkString.prototype.toJSON = RkString.prototype.valueOf = function () {
        return this.translatedValue;
    };
    function merge(to, from) {
        for (var i in from) {
            if (from.hasOwnProperty(i) && !to[i]) {
                to[i] = from[i];
            }
        }
        return to;
    }
    function getUrlWithoutParam(qeury, deleteParam) {
        if (Object.keys(qeury).length === 0 || Object.keys(qeury).length === 1 && qeury.hasOwnProperty(deleteParam)) {
            return '';
        }
        var result = '?';
        for (var name in qeury) {
            if (name !== deleteParam) {
                result += name + '=' + qeury[name] + '&';
            }
        }
        return result.slice(0, result.length - 1);
    }
    function init() {
        if (localizationEnabled) {
            i18n.setLang(i18n.detectLanguage());
        } else {
            i18n.setLang('');
        }
        if (!global.hasOwnProperty('rk')) {
            global.rk = i18n.rk.bind(i18n);
        }
    }
    i18n = {
        init: function () {
            if (this.isInit()) {
                return;
            }
            this.__init = true;
            init();
        },
        isInit: function () {
            return !!this.__init;
        },
        isEnabled: function () {
            return localizationEnabled;
        },
        setEnable: function (enable) {
            localizationEnabled = enable;
            init();
        },
        detectLanguage: function () {
            if (Util_1.constants.isServerScript) {
                return '';
            }
            if (Util_1.constants.isNodePlatform) {
                var detectedLng = Util_1.constants.defaultLanguage;
                var request = process.domain && process.domain.req;
                if (request) {
                    var reqCookie = request.cookies && request.cookies.lang;
                    var queryLang = request.query && request.query.lang;
                    var respond = process.domain.res;
                    detectedLng = queryLang || reqCookie || detectedLng;
                    detectedLng = this.hasLang(detectedLng) ? detectedLng : this.getDefaultLang();
                    if (respond && !(respond.cookies && respond.cookies.hasOwnProperty('lang')) && queryLang) {
                        Util_1.cookie.set('lang', detectedLng, {
                            expires: EXPIRES_COOKIES,
                            path: '/'
                        });
                        var redirectUrl = request.path + getUrlWithoutParam(request.query, 'lang');
                        respond.redirect(redirectUrl);
                    }
                }
                return detectedLng;
            }
            if (localizationEnabled) {
                var avLang = this.getAvailableLang();
                var detectedLng = Util_1.cookie.get('lang') || '';
                if (!detectedLng) {
                    detectedLng = Util_1.constants.defaultLanguage || global.contents && global.contents.defaultLanguage;
                }
                if (!detectedLng || detectedLng.length !== 5 || !avLang[detectedLng]) {
                    detectedLng = Object.keys(avLang)[0] || '';
                }
                return detectedLng;
            }
            return '';
        },
        getDefaultLang: function () {
            return Util_1.constants.defaultLanguage || 'ru-RU';
        },
        getLang: function () {
            if (Util_1.constants.isServerScript) {
                return '';
            }
            if (Util_1.constants.isNodePlatform) {
                return this.detectLanguage();
            }
            if (localizationEnabled) {
                return this._currentLang;
            }
            return '';
        },
        getAvailableLang: function () {
            return availableLanguage;
        },
        hasLang: function (language) {
            return language in availableLanguage;
        },
        setLang: function (language) {
            var _this = this;
            if (Util_1.constants.isServerScript || Util_1.constants.isNodePlatform) {
                return false;
            }
            if (localizationEnabled) {
                var changeLang = false, oldLang_1 = this._currentLang, currentLang_1;
                if (language && typeof language === 'string' && /..-../.test(language) && language !== this._currentLang) {
                    var parts = language.split('-');
                    this._currentLang = parts[0].toLowerCase() + '-' + parts[1].toUpperCase();
                    changeLang = true;
                }
                if (!language) {
                    this._currentLang = '';
                    changeLang = true;
                }
                if (changeLang) {
                    Util_1.cookie.set('lang', this._currentLang || null, {
                        expires: EXPIRES_COOKIES,
                        path: '/'
                    });
                    currentLang_1 = this._currentLang;
                    document.addEventListener('DOMContentLoaded', function () {
                        if (document.body.classList.length && oldLang_1) {
                            document.body.classList.remove(oldLang_1);
                        }
                        document.body.classList.add(currentLang_1);
                    });
                }
                return changeLang;
            }
            Util_1.cookie.set('lang', null, { path: '/' });
            document.addEventListener('DOMContentLoaded', function () {
                if (document.body.classList.length && _this._currentLang) {
                    document.body.classList.remove(_this._currentLang);
                }
            });
            return false;
        },
        _translate: function (key, ctx, num) {
            if (key === null || key === undefined || !key.indexOf) {
                return key;
            }
            var retValue = key;
            var lang = '';
            var index = key.indexOf(this._separator);
            if (index > -1) {
                ctx = key.substr(0, index);
                key = key.substr(index + this._separator.length);
            }
            if (typeof ctx === 'number') {
                num = ctx;
                ctx = '';
            }
            retValue = key;
            if (!Util_1.constants.isServerScript && (Util_1.constants.isNodePlatform || localizationEnabled)) {
                lang = this.getLang();
                if (lang && this._dict[lang]) {
                    if (num !== undefined) {
                        var trans_key = this._getTransKey(PLURAL_PREFIX + key, ctx, lang);
                        retValue = trans_key ? this._plural(trans_key, num) : key;
                    } else {
                        retValue = this._getTransKey(key, ctx, lang) || key;
                    }
                }
            }
            return retValue;
        },
        rk: function (key, ctx, num) {
            var _this = this;
            if (typeof window !== 'undefined' || key === null || key === undefined || !key.indexOf) {
                return this._translate(key, ctx, num);
            }
            return new RkString(key, function () {
                return _this._translate(key, ctx, num);
            });
        },
        _getTransKey: function (key, ctx, lang) {
            var trans_key = this._dict[lang][ctx ? '' + ctx + this._separator + key : '' + key];
            if (trans_key !== undefined) {
                return trans_key;
            }
            if (lang !== 'ru-RU' && typeof window !== 'undefined' && window.location.host.indexOf('-') > -1) {
                if (/[А-Яа-яA-Za-z]+/.test(key)) {
                    Util_1.IoC.resolve('ILogger').error('Localization', 'Для ключа ' + key + ' отсутствует перевод в словаре.');
                }
            }
            return undefined;
        },
        hasDict: function (dictName, lang) {
            lang = lang || this.getLang();
            return this._dictNames[lang] ? dictName in this._dictNames[lang] : false;
        },
        setDict: function (dict, name, lang) {
            lang = lang || this.getLang();
            if (lang && !this.hasDict(name, lang)) {
                if (name) {
                    this._dictNames[lang] = this._dictNames[lang] || {};
                    this._dictNames[lang][name] = true;
                }
                this._dict[lang] = merge(this._dict[lang] || {}, dict);
            }
        },
        getDictPath: function (moduleName, lang, ext) {
            var modulePath = Util_1.pathResolver.resolveComponentPath(moduleName), matched, firstLevelDir, dictPath, dictionary;
            dictionary = Util_1.constants.dictionary || {};
            if (Object.keys(dictionary).length === 0) {
                if (typeof global.contents !== 'undefined') {
                    dictionary = global.contents.dictionary || {};
                } else if (Util_1.constants.isNodePlatform) {
                    var contents = process.domain && process.domain.service && process.domain.service.getContents();
                    dictionary = contents && contents.dictionary || {};
                }
            }
            var resolveByRequire = false;
            if (!modulePath) {
                resolveByRequire = true;
                modulePath = moduleName;
            } else if (modulePath.indexOf('/') === 0) {
                resolveByRequire = true;
                modulePath = modulePath.slice(1);
            }
            if (resolveByRequire) {
                if (modulePath.substr(-ext.length - 1).indexOf('.' + ext) !== 0) {
                    modulePath = modulePath + '.' + ext;
                }
                modulePath = global.requirejs.toUrl(modulePath);
            }
            modulePath = modulePath.replace(dblSlashes, '/');
            matched = modulePath.match(directoryRegexp);
            firstLevelDir = matched && (matched[3] || matched[2]);
            if (firstLevelDir && dictionary[firstLevelDir + '.' + lang + '.' + ext]) {
                if (firstLevelDir === 'ws') {
                    firstLevelDir = 'WS';
                }
                dictPath = firstLevelDir + '/lang/' + lang + '/' + lang + '.' + ext;
            }
            return dictPath;
        },
        _plural: function (str, num) {
            if (str !== undefined) {
                num = Math.abs(num);
                var lang = this.getLang(), arg = void 0;
                arg = [num].concat(str.split(PLURAL_DELIMITER));
                switch (lang) {
                case 'en-US':
                    return this._pluralEn.apply(this, arg);
                case 'ru-RU':
                    return this._pluralRu.apply(this, arg);
                default:
                    return str;
                }
            }
            return undefined;
        },
        _pluralRu: function (num, word1, word2, word3, word4) {
            if (num % 1 > 0) {
                return word4;
            }
            num = num % 100;
            if (num >= 11 && num <= 19) {
                return word3;
            }
            num = num % 10;
            if (num == 1) {
                return word1;
            }
            if (num == 2 || num == 3 || num == 4) {
                return word2;
            }
            return word3;
        },
        _pluralEn: function (num, word1, word2) {
            if (num > 1 || num === 0) {
                return word2;
            }
            return word1;
        },
        __init: false,
        _separator: '@@',
        _currentLang: '',
        _dict: {},
        _dictNames: {},
        _rkString: RkString
    };
    i18n.init();
    exports.default = i18n;
});
