// sbis3-ws/Core/_Util/Compatibility.js
define('Core/_Util/Compatibility', [
    'require',
    'exports',
    'Core/_Util/BomDefine'
], function (require, exports) {
    'use strict';
    Object.defineProperty(exports, '__esModule', { value: true });
    var div = document && document.createElement('div');
    var divStyle = div && div.style;
    var cssTransformProperty = document && (div.style.transform !== undefined && 'transform' || div.style.msTransform !== undefined && 'msTransform' || div.style.oTransform !== undefined && 'oTransform' || div.style.MozTransform !== undefined && 'MozTransform' || div.style.webkitTransform !== undefined && 'webkitTransform');
    var isSupportPassive = function () {
        var supportsPassive = false;
        try {
            var opts = Object.defineProperty({}, 'passive', {
                get: function () {
                    supportsPassive = true;
                }
            });
            window.addEventListener('test', null, opts);
        } catch (e) {
        }
        return supportsPassive;
    };
    var input = document && document.createElement('input');
    var style = document && document.createElement('style');
    input && input.setAttribute('type', 'file');
    style && style.setAttribute('type', 'text/css');
    exports.default = {
        dateBug: new Date(2014, 0, 1, 0, 0, 0, 0).getHours() !== 0,
        placeholder: document && !!('placeholder' in input),
        fileupload: document && !input.disabled,
        correctKeyEvents: window && (!window.opera || parseFloat(window.opera.version()) >= 12.1),
        cssAnimations: document && (divStyle.transition !== undefined || divStyle.mozTransition !== undefined || divStyle.webkitTransition !== undefined),
        fullScreen: document && (div.requestFullscreen !== undefined || div.mozRequestFullScreen !== undefined || div.webkitRequestFullscreen !== undefined),
        wheel: document && (div.onwheel !== undefined || document.documentMode >= 9 ? 'wheel' : div.onmousewheel !== undefined ? 'mousewheel' : 'DOMMouseScroll'),
        cssTransformProperty: cssTransformProperty,
        cssTransform: !!cssTransformProperty,
        touch: document && (navigator.msMaxTouchPoints || navigator.maxTouchPoints || 'ontouchstart' in document.documentElement),
        standartStylesheetProperty: document && !style.styleSheet,
        supportPassive: isSupportPassive()
    };
});
