// sbis3-ws/Core/_Util/Constants.js
define('Core/_Util/Constants', [
    'require',
    'exports',
    'Core/_Util/Compatibility',
    'Core/_Util/Detection'
], function (require, exports, Compatibility_1, Detection_1) {
    'use strict';
    Object.defineProperty(exports, '__esModule', { value: true });
    var now = +new Date();
    var timeShift2014 = +new Date(2014, 9, 26);
    var constants = {
        JSONRPC_PROOTOCOL_VERSION: 5,
        appRoot: '/',
        resourceRoot: '/resources/',
        wsRoot: '/ws/',
        cdnRoot: '/cdn/',
        defaultServiceUrl: '/service/',
        logLevel: 'info',
        buildnumber: '',
        isBrowserPlatform: typeof window !== 'undefined',
        isNodePlatform: typeof process !== 'undefined',
        isServerScript: typeof window === 'undefined' && typeof process === 'undefined',
        isBuildOnServer: typeof process !== 'undefined' && process.application,
        isNode: typeof process !== 'undefined' && process && process.release && process.release.name === 'node',
        compat: (typeof window === 'undefined' || window && window.location.href.indexOf('withoutLayout') === -1) && (typeof window === 'undefined' || window.wsConfig.compatible !== false),
        i18n: false,
        defaultLanguage: '',
        availableLanguage: {},
        dictionary: {},
        theme: 'wi_scheme',
        tensorFontsUrl: [
            'Controls-theme/themes/default/fonts/TensorFont.css',
            'Controls-theme/themes/default/fonts/cbus-icons.css',
            'Controls-theme/themes/default/helpers/general.css'
        ],
        hosts: [],
        services: {},
        modules: {},
        jsCoreModules: {},
        jsModules: {},
        compatibility: Compatibility_1.default,
        browser: Detection_1.default,
        moscowTimeOffset: now >= timeShift2014 ? 3 * 60 : 4 * 60,
        styleLoadTimeout: 4000,
        IDENTITY_SPLITTER: ',',
        userConfigSupport: false,
        globalConfigSupport: true,
        debug: false,
        saveLastState: true,
        checkSessionCookie: true,
        layoutConfig: {
            contentArea: [
                '#contentArea',
                '#contentAreaEdit',
                '#ctr'
            ]
        },
        isNewNav: true,
        xmlPackages: {},
        jsPackages: {},
        xmlContents: {},
        htmlNames: {},
        hdlBindings: {},
        startDragDistance: 4,
        WINDOW_CLASSES: [
            'Lib/Control/FloatArea/FloatArea',
            'Lib/Control/Window/Window',
            'Deprecated/Controls/FieldEditAtPlace/FieldEditAtPlace'
        ],
        key: {
            left: 37,
            up: 38,
            right: 39,
            down: 40,
            insert: 45,
            del: 46,
            space: 32,
            backspace: 8,
            minus: 109,
            plus: 107,
            enter: 13,
            esc: 27,
            f1: 112,
            f3: 114,
            f4: 115,
            f5: 116,
            f7: 118,
            f12: 123,
            meta: 157,
            underscore: 95,
            pageUp: 33,
            pageDown: 34,
            end: 35,
            home: 36,
            tab: 9,
            ctrl: 17,
            b: 66,
            h: 72,
            v: 86,
            y: 89,
            q: 81,
            p: 80,
            m: 77,
            n: 78,
            o: 79,
            androidAnyKey: 229
        },
        modifiers: {
            nothing: 0,
            shift: 2,
            control: 4,
            alt: 8
        },
        operaKeys: {
            43: 'plus',
            45: 'minus'
        },
        NON_CTRL: '.ws-non-control',
        fasttemplate: false,
        nostyle: false,
        javaStartTimeout: 15000,
        resizeDelay: 40,
        classLoaderMappings: {},
        IE_ACTIVEOBJECT_XML_PRINT_TYPE: 'Microsoft.XmlDom'
    };
    var global = function () {
        return this || (0, eval)('this');
    }();
    global.IS_PRODUCTION = constants.isProduction;
    exports.default = constants;
});
