// sbis3-ws/Core/_Util/IoC.js
define('Core/_Util/IoC', [
    'require',
    'exports'
], function (require, exports) {
    'use strict';
    Object.defineProperty(exports, '__esModule', { value: true });
    var map = {}, singletons = {};
    function resolveAsFunction(func, config) {
        var result;
        if (func instanceof Function && func.prototype && func.prototype.$constructor) {
            result = new func(config);
        } else {
            result = func(config);
        }
        return result;
    }
    exports.default = {
        bind: function (interfaceName, implementationName) {
            map[interfaceName] = {
                implementation: implementationName,
                isSingle: 0
            };
        },
        bindSingle: function (interfaceName, implementationName, config) {
            map[interfaceName] = {
                implementation: implementationName,
                isSingle: 1,
                config: config || {}
            };
            singletons[interfaceName] = '';
        },
        resolve: function (interfaceName, config) {
            if (interfaceName in map) {
                var binding = map[interfaceName], classConstructorName = binding.implementation, isSingleton = binding.isSingle, implementation = void 0;
                if (isSingleton && singletons[interfaceName]) {
                    return singletons[interfaceName];
                }
                if (typeof classConstructorName === 'function') {
                    implementation = classConstructorName;
                } else if (typeof window === 'undefined' && require.defined('Core/core-structure')) {
                    var $ws = require('Core/core-structure');
                    if (typeof $ws.proto[classConstructorName] === 'function') {
                        implementation = $ws.proto[classConstructorName];
                    }
                }
                if (implementation) {
                    if (isSingleton) {
                        return singletons[interfaceName] = resolveAsFunction(implementation, binding.config);
                    }
                    return resolveAsFunction(implementation, config);
                }
                return binding.implementation;
            }
            throw new ReferenceError('No mappings defined for ' + interfaceName);
        },
        has: function (interfaceName) {
            return interfaceName in map;
        }
    };
});
